nexus/types/src/deployment/network_resources.rs

```diff
@@ -0,0 +1,283 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+use super::tri_map::TriMap;
+use super::tri_map::TriMapEntry;
+use anyhow::anyhow;
+use omicron_common::api::external::MacAddr;
+use omicron_common::api::internal::shared::SourceNatConfig;
+use omicron_uuid_kinds::ExternalIpUuid;
+use omicron_uuid_kinds::OmicronZoneUuid;
+use omicron_uuid_kinds::VnicUuid;
+use schemars::JsonSchema;
+use serde::Deserialize;
+use serde::Serialize;
+use std::net::IpAddr;
+use std::net::SocketAddr;
+use thiserror::Error;
+
+/// Tracker and validator for network resources allocated to Omicron-managed zones.
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct OmicronZoneNetworkResources {
+    /// external IPs allocated to Omicron zones
+    omicron_zone_external_ips: TriMap<OmicronZoneExternalIpEntry>,
+
+    /// vNICs allocated to Omicron zones
+    omicron_zone_nics: TriMap<OmicronZoneNicEntry>,
```

Comment:
Would it make sense to squish these two maps together? I ended up combining external IP + NIC in the [BlueprintZoneType helper method](https://github.com/oxidecomputer/omicron/blob/bca1f2d63b70e71aef509b94e190e54c6eb3a6be/nexus/types/src/deployment/zone_type.rs#L36-L38) because all zone types either have both or neither. (Now that I ask that, that might require a 5-way map instead of a 3-way map? :grimacing:)

Reply:
Sure, I think this pattern scales up to any N.

Reply:
Chatted with @jgallagher a bit more. While it is true that these are 1:1 in the blueprint:

* that might not be the case in the database itself
* within the planning input we're not depending on the 1:1 property across IPs and NICs right now
* the future need for this map is for expunged zone garbage collection. That doesn't require 1:1 across IPs and NICs.

So based on all this I'll stick with the pair of tri-maps and add a comment explaining this.

Reply:
Added a comment.